File: src/rl_agent/cointegration.R
# An Reinforcement Learning Application for Statistical Arbitrage
# karol.przybylak.13@ucl.ac.uk, kmprzybylak@gmail.com

# Cointegrations


# -----------------------------------------------------------------
#region Dependencies

# Cointegration
p_load(urca, tseries)

#endregion
# -----------------------------------------------------------------
#region Function Definitions


# GetCOintPairs() inputs a data frame with timeseries data in columns and
# returns a data frame where each column includes the names of cointegrated pairs

# uwaga: w notatkach z StackExchange widac, ze moze nam to rownoczesnie wykryc stacjonarne szeregi. Moze uzyc tego?

GetCointPairs <- function(df_currencies, c_currencies, learn_period, cval) { 
  
  # returns a df where each column includes names of the cointegrated series based on data df_currencies
  # cval (1,2,3) is the critical value - (10 percentile,5pct,1pct)
  
  # Cut the data to the specified learn period
  df_currencies <- df_currencies[c(learn_period[1]:learn_period[2]),]
  # all 2-element combinations of the given currency pair names
  df_combn <- combn(c_currencies, 2) 
  rejected_combn <- numeric()
  assert_that(NCOL(df_combn) > 0, msg = 'df_combn can\'t be empty')
  # find the ones that are not cointegrated with the specified p-value  
  for( i in 1 : NCOL(df_combn)) {
    
    if( ca.jo(df_currencies[df_combn[,i]])@teststat[2] < ca.jo(df_currencies[df_combn[,i]])@cval[2,cval]) { 
      # jezeli odrzucamy na poziomi 0.95 to skreslamy kombinacj? par z dalszych rozwazan//wypisac df_curr jako argument? najpierw trzeba wywolac df currencies. head(df_currencies[combn[,1]]) dzia?a.
      rejected_combn <- c(rejected_combn, i) #colnames(df_combn) # saves columns of uncointegrated pairs
    } 
  }
  df_combn %<>% as.data.frame() %>% select(-rejected_combn)
    #subset w ponizszej funkcji zle dziala przy pustym wektorze w subset dlatego takie IF...
  return(df_combn)
}

CointSpreads <- function(coint_pairs, df_currencies, learn_period, test_period){
  
  # returns the cointegration spread for pairs of known cointegrated series
  # devides this spread into train/test, where test 
  # has the cointegration coefficients from the train period
   
  l_sets <- list()
  
  df_train <- df_currencies[c(learn_period[1]: learn_period[2]),]
  df_test <- df_currencies[c(test_period[1]: test_period[2]),]
  df_all <- df_currencies
  
  for( i in (1 : (ncol(coint_pairs)))){
    pair <- as.character(coint_pairs[, i])
    
    train_sample <- df_train[,pair]
    test_sample <- df_test[, pair]
    all <- df_all[, pair]
    
    coint_vec <- ca.jo(train_sample)@V[,1]    # coint coefs from the learning period / cointegration coefficients for the given pair in the first period / wouldn' it be better to return this already in GetCointPairs() ?
    train_spread <- coint_vec[1] * train_sample[, 1] + coint_vec[2] * train_sample[, 2]
    test_spread <-  coint_vec[1] * test_sample[, 1] + coint_vec[2] * test_sample[, 2]
    all_spread <-   coint_vec[1] * df_all[, 1] + coint_vec[2] * df_all[, 2]
    
    l_sets[["train"]][[paste0(coint_pairs[1,i],"+", coint_pairs[2, i])]] <- train_spread
    l_sets[["test"]][[paste0(coint_pairs[1,i], "+", coint_pairs[2, i])]] <- test_spread
    l_sets[["all"]][[paste0(coint_pairs[1,i],  "+", coint_pairs[2, i])]] <- all_spread
    l_sets[["coef"]][[paste0(coint_pairs[1,i], "+", coint_pairs[2, i])]] <- coint_vec
  }    
  
  return(l_sets)
  
}

GetPairs <- function(df_currencies, c_currencies, learn_period){
  
  df_train <- df_currencies[c(learn_period[1]: learn_period[2]),]
  df_test  <- df_currencies[c(test_period[1]: test_period[2]),]
  df_all   <- df_currencies
  l_sets   <- alply(c_currencies, 1, function(cur){
    return(list(train = df_train[, cur],
                test  = df_test[,  cur],
                all   = df_all[,   cur]))
  })
  names(l_sets) <- c_currencies
  return(l_sets)
}

File: src/rl_agent_python/cointegration.py
"""
Cointegration analysis for statistical arbitrage.

This module provides functions for finding cointegrated pairs of assets
and calculating their spreads.
"""

from typing import Dict, List, Tuple

import pandas as pd
from statsmodels.tsa.vector_ar.vecm import coint_johansen


def get_cointegrated_pairs(
    df_currencies: pd.DataFrame,
    c_currencies: List[str],
    learn_period: Tuple[int, int],
    critical_value: int = 2,
) -> pd.DataFrame:
    """
    Find cointegrated pairs of assets.

    Args:
        df_currencies: DataFrame with time series data in columns
        c_currencies: List of currency/asset names to check for cointegration
        learn_period: Tuple of (start, end) indices for the learning period
        critical_value: Critical value index (0: 10%, 1: 5%, 2: 1%)

    Returns:
        DataFrame where each column contains names of cointegrated pairs
    """
    # Cut the data to the specified learn period
    df_currencies = df_currencies.iloc[learn_period[0] : learn_period[1]]

    # Get all 2-element combinations of the given currency pair names
    from itertools import combinations

    pairs = list(combinations(c_currencies, 2))

    if not pairs:
        raise ValueError("No currency pairs to check for cointegration")

    cointegrated_pairs = []

    # Find pairs that are cointegrated with the specified p-value
    for pair in pairs:
        # Perform Johansen cointegration test
        result = coint_johansen(df_currencies[pair], det_order=0, k_ar_diff=1)

        # Check if the test statistic exceeds the critical value
        if result.lr1[0] > result.cvt[0, critical_value]:
            cointegrated_pairs.append(pair)

    # Convert to DataFrame format similar to R version
    if cointegrated_pairs:
        df_pairs = pd.DataFrame(cointegrated_pairs).T
        df_pairs.columns = [f"V{i + 1}" for i in range(len(cointegrated_pairs))]
    else:
        df_pairs = pd.DataFrame(columns=["V1"])

    return df_pairs


def cointegration_spreads(
    coint_pairs: pd.DataFrame,
    df_currencies: pd.DataFrame,
    learn_period: Tuple[int, int],
    test_period: Tuple[int, int],
) -> Dict:
    """
    Calculate cointegration spreads for pairs of known cointegrated series.

    Args:
        coint_pairs: DataFrame containing cointegrated pairs
        df_currencies: DataFrame with time series data
        learn_period: Tuple of (start, end) indices for the learning period
        test_period: Tuple of (start, end) indices for the test period

    Returns:
        Dictionary containing train, test, and all spreads, plus coefficients
    """
    l_sets = {
        "train": {},
        "test": {},
        "all": {},
        "coef": {},
    }

    df_train = df_currencies.iloc[learn_period[0] : learn_period[1]]
    df_test = df_currencies.iloc[test_period[0] : test_period[1]]
    df_all = df_currencies

    for col in coint_pairs.columns:
        pair = coint_pairs[col].tolist()
        pair_name = f"{pair[0]}+{pair[1]}"

        # Get cointegration coefficients from the learning period
        result = coint_johansen(df_train[pair], det_order=0, k_ar_diff=1)
        coint_vec = result.evec[:, 0]  # First eigenvector

        # Calculate spreads
        train_spread = (
            coint_vec[0] * df_train[pair[0]] + coint_vec[1] * df_train[pair[1]]
        )
        test_spread = coint_vec[0] * df_test[pair[0]] + coint_vec[1] * df_test[pair[1]]
        all_spread = coint_vec[0] * df_all[pair[0]] + coint_vec[1] * df_all[pair[1]]

        # Store results
        l_sets["train"][pair_name] = train_spread
        l_sets["test"][pair_name] = test_spread
        l_sets["all"][pair_name] = all_spread
        l_sets["coef"][pair_name] = coint_vec

    return l_sets


def get_pairs(
    df_currencies: pd.DataFrame,
    c_currencies: List[str],
    learn_period: Tuple[int, int],
    test_period: Tuple[int, int],
) -> Dict:
    """
    Get train/test/all data for individual currencies.

    Args:
        df_currencies: DataFrame with time series data
        c_currencies: List of currency names
        learn_period: Tuple of (start, end) indices for the learning period
        test_period: Tuple of (start, end) indices for the test period

    Returns:
        Dictionary containing train, test, and all data for each currency
    """
    df_train = df_currencies.iloc[learn_period[0] : learn_period[1]]
    df_test = df_currencies.iloc[test_period[0] : test_period[1]]
    df_all = df_currencies

    l_sets = {}
    for currency in c_currencies:
        l_sets[currency] = {
            "train": df_train[currency],
            "test": df_test[currency],
            "all": df_all[currency],
        }

    return l_sets
